.github/workflows/ci.yml

```diff
@@ -90,37 +75,53 @@ jobs:
       env:
         COLOR: 'yes'
       run: |
-        make test-only
+        make test
         python -m coverage xml
     - name: Upload coverage
       uses: codecov/codecov-action@v3
       with:
         file: ./coverage.xml
         flags: unit
-        fail_ci_if_error: false
+
+  check:  # This job does nothing and is only used for the branch protection
+    if: always()
+
+    needs: [lint, test]
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Decide whether the needed jobs succeeded or failed
+      uses: re-actors/alls-green@release/v1
+      with:
+        jobs: ${{ toJSON(needs) }}
 
   deploy:
     name: Deploy
+    environment: release
     runs-on: ubuntu-latest
-    needs: test
-    # Run only on pushing a tag
+    needs: [check]
     if: github.event_name == 'push' && contains(github.ref, 'refs/tags/')
     steps:
     - name: Checkout
       uses: actions/checkout@v3
-    - name: Setup Python 3.8
+    - name: Setup Python
       uses: actions/setup-python@v4
       with:
         python-version: 3.8
     - name: Install dependencies
       run:
-        python -m pip install -U pip wheel twine build
-    - name: Make dists
-      run:
-        python -m build
-    - name: PyPI upload
-      env:
-        TWINE_USERNAME: __token__
-        TWINE_PASSWORD: ${{ secrets.PYPI_TOKEN }}
+        python -m pip install -U pip wheel setuptools build twine
+    - name: Build dists
       run: |
-        twine upload dist/*
+        python -m build
```

Comment:
@Dreamsorcerer FYI it's best to build in a separate job to avoid exposing tokens to arbitrary code. 

I've actually been encouraging aiomysql and aiomonitor folks use my new (kinda) thing — https://github.com/marketplace/actions/checkout-python-sdist. It lets one run tests right from an sdist.

https://github.com/aio-libs/aiomonitor/blob/main/.github/workflows/ci-cd.yml and https://github.com/aio-libs/aiomysql/blob/master/.github/workflows/ci-cd.yml are closer to what I want for aio-libs CI in general. Hopefully, later I'll be able to standardize things more, through reuse, in particular.

Reply:
Yep, I've been thinking about ways to standardise some of our configuration too, so we can update everything in one place.
I noticed that ruff has an option to include a config file from another config file, so I was wondering about migrating from flake8 to ruff, with a base config in a dedicated config repo. That could then be included into our other repos with a gitmodule (and a repo-specific config can override the base options), which can then be automatically updated with dependabot.

Reply:
Yeah, I've been looking into ruff, but not quite ready to migrate. I'd like to try migrating all the rules that are implemented there, but still run flake8 for things that are coming from non-migrated plugins.

As I wrote in another comment, I'd go for a merging in a git [skeleton](https://github.com/jaraco/skeleton) repo rather than using submodules — they can really complicate the DX. I want to have our own base git skeleton repository version that we'd be merging into the in-org project. Maybe a few of them, with different sets of configs. And in addition to this, we'd need a separate repository with some reusable workflows for most of CI configuration, excluding the secretless publishing to PyPI since it currently doesn't support those. As those workflow files would be integrated into the git skeleton, all the projects would get them. The workflows would have to be versioned similarly to composite actions, for stability.

It'd also be nice to integrate a https://nitpick.rtfd.io.

There are other things to do, though. For example, I'd like to get to making reusable multiple per-env lockfile generation, like what I was experimenting with in the Cheroot project. It doesn't have the best DX, though — it ended up polluting Git history excessively, so I'll need to invent a way of reducing that noise (I do have some ideas). It would need to replace dependabot for the requirements files. It's quite noisy anyway.